src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -0,0 +1,1880 @@
+            dAi(freq,i) = (float) i / (m1sizeX * sampling_rate);
+
+			if (i>0)
+			{
+				id=mdRes.addObject();
+				mdRes.setValue(MDL_RESOLUTION_FREQ,dAi(freq, i),id);
+				mdRes.setValue(MDL_RESOLUTION_FRC,dAi(frc, i),id);
+				mdRes.setValue(MDL_RESOLUTION_FREQREAL, 1./dAi(freq, i), id);
+			}
+        }
+
+        FOR_ALL_ELEMENTS_IN_ARRAY1D(freq)
+		 {
+			 if ( (dAi(frc,i)<=thrs) && (i>2) )
+			 {
+				double y2, y1, x2, x1, slope, ny;
+				y2 = dAi(freq,i);
+				y1 = dAi(freq,i-1);
+				x2 = dAi(frc,i);
+				x1 = dAi(frc,i-1);
+
+				slope = (y2 - y1)/(x2 - x1);
+				ny = y2 - slope*x2;
+
+				resInterp = 1/(slope*thrs + ny);
+				 fscFreq = 1.0/dAi(freq, i);
+				 break;
+			 }
+		 }
+
+       	mdRes.write(fn_fscmd_folder+"GlobalFSC.xmd");
+
+        std::cout << "    " << std::endl;
+    }
+
+
+
+    void createfrequencySphere(MultidimArray<double> &sphere,
+    		Matrix1D<double> &freq_fourier_x,
+			 Matrix1D<double> &freq_fourier_y,
+			 Matrix1D<double> &freq_fourier_z)
+    {
+    	int ZdimFT1=(int)ZSIZE(sphere);
+		int YdimFT1=(int)YSIZE(sphere);
+		int XdimFT1=(int)XSIZE(sphere);
+
+    	long n = 0;
+    	sphere.initConstant(-0.5);
+		for (int k=0; k<ZdimFT1; k++)
+		{
+			double uz = VEC_ELEM(freq_fourier_z,k);
+			uz *= uz;
+			for (int i=0; i<YdimFT1; i++)
+			{
+				double uy = VEC_ELEM(freq_fourier_y,i);
+				uy *= uy;
+				for (int j=0; j<XdimFT1; j++)
+				{
+					double ux = VEC_ELEM(freq_fourier_x,j);
+					ux *= ux;
+					ux = sqrt(ux + uy + uz);
+
+					if (ux>0.5)
+					{
+						++n;
+						continue;
+					}
+					else
+						DIRECT_MULTIDIM_ELEM(sphere,n) = -ux;
+					++n;
+				}
+			}
+		}
+    }
+
+
+    void crossValues(Matrix2D<double> &indexesFourier, double &rot, double &tilt, double &angCon,
+			 MultidimArray<std::complex<double>> &f1, MultidimArray<std::complex<double>> &f2,
+			 std::complex<double> &f1_mean, std::complex<double> &f2_mean)
+    {
+    	double x_dir, y_dir, z_dir, cosAngle, aux;
+		double lastCosine = 0;
+
+		x_dir = sin(tilt*PI/180)*cos(rot*PI/180);
+		y_dir = sin(tilt*PI/180)*sin(rot*PI/180);
+		z_dir = cos(tilt*PI/180);
+
+		cosAngle = cos(angCon*PI/180);
+		aux = 4.0/((cos(angCon*PI/180) -1)*(cos(angCon*PI/180) -1));
+
+		double counter_ = 0;
+		double wt = 0;
+
+		long n = 0;
+		FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(f1)
+		{
+			double cosine = x_dir*MAT_ELEM(indexesFourier, 1, n) + y_dir*MAT_ELEM(indexesFourier, 2, n) +
+							z_dir*MAT_ELEM(indexesFourier, 0, n);
+
+			cosine = fabs(cosine);
+
+			if (cosine>=cosAngle)
+			{
+//        		std::cout << "wt = " << sqrt(exp( -((cosine -1)*(cosine -1))*aux )) << std::endl;
+				wt += sqrt(exp( -((cosine -1)*(cosine -1))*aux ));
+//        		std::cout << "wt = " << wt << std::endl;
+			}
+		}
+		wt = 1/wt;
+//        std::cout << "-----------" << std::endl;
+//        std::cout << "wt" << wt << std::endl;
+//        std::cout << "-----------" << std::endl;
+
+		std::complex<double> f1_orig, f2_orig;
+		f1_mean = (0, 0);
```

Comment:
See https://stackoverflow.com/a/36169624/5484355 (TLDR: use ideally `auto f1_orig = std::complex<double>(0, 0);`)

Reply:
I think it's done